hackathon_site/hardware/serializers.py

```diff
@@ -34,9 +38,131 @@ def get_unique_hardware_count(obj: Category):
+        # "id": <Hardware Object>, and
+        # "quantity": <Int>
+        return functools.reduce(
+            lambda x, y: x + y,
+            [Counter({e["id"]: e["quantity"]}) for e in hardware_requests],
+        )
+
+    # check that the requests are within per-team constraints
+    def validate(self, data):
+        # requested_hardware is a Counter where the keys are <Hardware Object>'s and values are <Int>'s
+        requested_hardware = self.merge_requests(hardware_requests=data["hardware"])
+        hardware_query = (
+            Hardware.objects.filter(
+                id__in=[hardware.id for hardware in requested_hardware.keys()],
+            )
+            .all()
+            .prefetch_related("categories", "order_items")
+        )
+        team_unreturned_orders = hardware_query.annotate(
+            past_order_count=Count(
+                "order_items",
+                filter=Q(order_items__part_returned_health__isnull=True)
+                & ~Q(order_items__order__status="Cancelled")
+                & Q(order_items__order__team=self.context["request"].user.profile.team),
+                distinct=True,
+            )
+        )
+        category_counts = dict()
+        error_messages = []
+        for (hardware, requested_quantity) in requested_hardware.items():
+            team_hardware = team_unreturned_orders.get(id=hardware.id)
+            team_hardware_count = getattr(team_hardware, "past_order_count", 0)
+            if hardware.max_per_team < (team_hardware_count + requested_quantity):
+                error_messages.append("Hardware {} limit reached".format(hardware.name))
+            for category in hardware.categories.all():
+                category_counts[category] = (
+                    category_counts.get(category, 0)
+                    + team_hardware_count
+                    + requested_quantity
+                )
+        for (category, count) in category_counts.items():
+            if category.max_per_team < count:
+                error_messages.append("Category {} limit reached".format(category.name))
+        if error_messages:
+            error_message = "; ".join(error_messages)
+            raise serializers.ValidationError(error_message)
+        return data
+
+    def create(self, validated_data):
+        # validated data should already satisfy all constraints
+        requested_hardware = self.merge_requests(
+            hardware_requests=validated_data["hardware"]
+        )
+        new_order = None
+        response_data = {"order_id": -1, "hardware": [], "errors": []}
+        for (hardware, requested_quantity) in requested_hardware.items():
+            num_order_items = min(hardware.quantity_remaining, requested_quantity)
+            if num_order_items <= 0:
+                response_data["hardware"].append({hardware.id: 0})
+                response_data["errors"].append(
+                    {hardware.id: "There are no {} available".format(hardware.name)}
```

Comment:
Minor, can make this `There are no {}s available` for grammar in most cases

Reply:
done :)